### PR TITLE
feat: adds universe domain for DownscopedCredentials and ExternalAccountAuthorizedUserCredentials

### PR DESCRIPTION
**Pull request by lsirac**
- Adds universe domain support for DownscopedCredentials.
- Adds universe domain support for ExternalAccountAuthorizedUserCredentials.
- Adds more tests for universe domain in the BYOID credential classes.

### CODE REVIEW

File: oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java [hunk 1]
```diff
@@ -87,22 +88,32 @@
...
-            transportFactory,
+            builder.transportFactory,
             getFromServiceLoader(HttpTransportFactory.class, OAuth2Utils.HTTP_TRANSPORT_FACTORY));
-    this.sourceCredential = checkNotNull(sourceCredential);
-    this.credentialAccessBoundary = checkNotNull(credentialAccessBoundary);
+    this.sourceCredential = checkNotNull(builder.sourceCredential);
+    this.credentialAccessBoundary = checkNotNull(builder.credentialAccessBoundary);
+
+    // Default to GDU when not supplied.
+    if (builder.universeDomain == null || builder.universeDomain.trim().isEmpty()) {
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java:110]: nit: should we check the source credentials universe domain? order would be builder -> source credential -> default

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java [hunk 2]
```diff
@@ -329,6 +330,7 @@ static ExternalAccountAuthorizedUserCredentials fromJson(
         .setRefreshToken(refreshToken)
         .setHttpTransportFactory(transportFactory)
         .setQuotaProjectId(quotaProjectId)
+        .setUniverseDomain(universeDomain)
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java:333]: nit: defaulting logic? Or is that handled in a parent class already

**Comment by TimurSadykov** [comment]
looking

File: oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java [hunk 3]
```diff
@@ -87,22 +88,32 @@
  */
 public final class DownscopedCredentials extends OAuth2Credentials {
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java:89]: Noticed this does not override toString, hashcode and equals. Is it expected?

File: oauth2_http/javatests/com/google/auth/oauth2/AwsCredentialsTest.java [hunk 4]
```diff
@@ -964,7 +965,44 @@ public void shouldUseMetadataServer_noEnvironmentVars() {
...
+    assertEquals(
+        SERVICE_ACCOUNT_IMPERSONATION_URL, credentials.getServiceAccountImpersonationUrl());
+    assertEquals(AWS_CREDENTIAL_SOURCE, credentials.getCredentialSource());
+    assertEquals("quotaProjectId", credentials.getQuotaProjectId());
+    assertEquals("clientId", credentials.getClientId());
+    assertEquals("clientSecret", credentials.getClientSecret());
+    assertEquals(scopes, credentials.getScopes());
+    assertEquals(SystemEnvironmentProvider.getInstance(), credentials.getEnvironmentProvider());
+    assertEquals("universeDomain", credentials.getUniverseDomain());
+  }
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/AwsCredentialsTest.java:1002]: add credentials.toBuilder().build case or a separate test? This one might be losing the awsCredentialSource. 
(I have found some bugs related to this in other creds when adding universe_domain.) This applies to all 3PI credentials

**Review by TimurSadykov — CHANGES_REQUESTED** [review]

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java (hunk 2 above)
Comment (lsirac) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java:333]: Yes, defaults to GDU in GoogleCredentials.

File: oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java (hunk 3 above)
Comment (lsirac) [reply oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java:89]: Not sure why we didn't add them. I opened a bug to track -- will add in another PR.

File: oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java [hunk 5]
```diff
@@ -112,6 +112,18 @@ private DownscopedCredentials(Builder builder) {
...
+
+    // Ensure source credential's universe domain matches.
+    try {
+      if (!this.universeDomain.equals(sourceCredential.getUniverseDomain())) {
+        throw new IllegalArgumentException(
+            "The downscoped credential's universe domain must be the same as the source "
+                + "credential.");
+      }
+    } catch (IOException e) {
+      throw new IllegalStateException(
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java:124]: nit: is it possible to add the IOException as the cause of the IllegalStateException here?

**Review by aeitzman — APPROVED** [review]

File: oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java [hunk 6]
```diff
@@ -87,22 +88,44 @@
...
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java:124]: comment on why is it handled like that?

File: oauth2_http/javatests/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentialsTest.java [hunk 7]
```diff
@@ -301,13 +333,35 @@ public void toBuilder() {
             .setRevokeUrl(REVOKE_URL)
             .setAccessToken(new AccessToken(ACCESS_TOKEN, new Date()))
             .setQuotaProjectId(QUOTA_PROJECT)
+            .setUniverseDomain(UNIVERSE_DOMAIN)
             .build();
 
     ExternalAccountAuthorizedUserCredentials secondCredentials = credentials.toBuilder().build();
 
     assertEquals(credentials, secondCredentials);
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentialsTest.java:341]: can you also add a test for universeDomain equals? like for other fields? equals_differentUniverseDomain()

This check base implementation is called

**Review by TimurSadykov — APPROVED** [review]
LGTM, one test suggestions

**Pull request merged**